prober/derp.go

```diff
@@ -0,0 +1,379 @@
+	if err != nil {
+		d.Lock()
+		defer d.Unlock()
+		if d.lastDERPMap != nil && time.Since(d.lastDERPMapAt) < 10*time.Minute {
+			log.Printf("Error while fetching DERP map, using cached one: %s", err)
+			// Assume that control is restarting and use
+			// the same one for a bit.
+			return nil
+		}
+		return err
+	}
+	defer res.Body.Close()
+	if res.StatusCode != 200 {
+		return fmt.Errorf("fetching %s: %s", d.derpMapURL, res.Status)
+	}
+	dm := new(tailcfg.DERPMap)
+	if err := json.NewDecoder(res.Body).Decode(dm); err != nil {
+		return fmt.Errorf("decoding %s JSON: %v", d.derpMapURL, err)
+	}
+
+	d.Lock()
+	defer d.Unlock()
+	d.lastDERPMap = dm
+	d.lastDERPMapAt = time.Now()
+	d.nodes = make(map[string]*tailcfg.DERPNode)
+	for _, reg := range d.lastDERPMap.Regions {
+		for _, n := range reg.Nodes {
+			if existing, ok := d.nodes[n.HostName]; ok {
+				return fmt.Errorf("derpmap has duplicate nodes: %+v and %+v", existing, n)
+			}
+			d.nodes[n.HostName] = n
+		}
+	}
+	return nil
+}
+
+func (d *derpProber) ProbeUDP(ipaddr string, port int) ProbeFunc {
+	return func(ctx context.Context) error {
+		_, err := derpProbeUDP(ctx, ipaddr, port)
+		return err
+	}
+}
+
+func derpProbeUDP(ctx context.Context, ipStr string, port int) (latency time.Duration, err error) {
+	pc, err := net.ListenPacket("udp", ":0")
+	if err != nil {
+		return 0, err
+	}
+	defer pc.Close()
+	uc := pc.(*net.UDPConn)
+
+	tx := stun.NewTxID()
+	req := stun.Request(tx)
+
+	if port == 0 {
+		port = 3478
+	}
+	for {
+		ip := net.ParseIP(ipStr)
+		_, err := uc.WriteToUDP(req, &net.UDPAddr{IP: ip, Port: port})
+		if err != nil {
+			return 0, err
+		}
+		buf := make([]byte, 1500)
+		uc.SetReadDeadline(time.Now().Add(2 * time.Second))
+		t0 := time.Now()
+		n, _, err := uc.ReadFromUDP(buf)
+		d := time.Since(t0)
+		if err != nil {
+			if ctx.Err() != nil {
+				return 0, fmt.Errorf("timeout reading from %v: %v", ip, err)
+			}
+			if d < time.Second {
+				return 0, fmt.Errorf("error reading from %v: %v", ip, err)
+			}
+			time.Sleep(100 * time.Millisecond)
+			continue
+		}
+		txBack, _, err := stun.ParseResponse(buf[:n])
+		if err != nil {
+			return 0, fmt.Errorf("parsing STUN response from %v: %v", ip, err)
+		}
+		if txBack != tx {
+			return 0, fmt.Errorf("read wrong tx back from %v", ip)
+		}
+		if latency == 0 || d < latency {
+			latency = d
+		}
+		break
+	}
+	return latency, nil
+}
+
+func derpProbeNodePair(ctx context.Context, dm *tailcfg.DERPMap, from, to *tailcfg.DERPNode) (latency time.Duration, err error) {
+	fromc, err := newConn(ctx, dm, from)
+	if err != nil {
+		return 0, err
+	}
+	defer fromc.Close()
+	toc, err := newConn(ctx, dm, to)
+	if err != nil {
+		return 0, err
+	}
+	defer toc.Close()
+
+	// Wait a bit for from's node to hear about to existing on the
+	// other node in the region, in the case where the two nodes
+	// are different.
+	if from.Name != to.Name {
+		time.Sleep(100 * time.Millisecond) // pretty arbitrary
```

Comment:
I don't quite understand what this 100 msec is, but some of our distant DERPs will be more than 100 msec away from wherever this prober runs. If this is RTT I think this will skip some of the DERPs.

Reply:
I believe this is to allow the information about a new node to propagate from the `to` DERP server to the `from` DERP server. Both DERP servers are in the same region, so 100ms seems sufficient?